test/Test/Pos/Block/Logic/Event.hs

```diff
@@ -77,15 +73,47 @@ runBlockEvent ::
        BlockLrcMode SscGodTossing BlockTestContext m
     => BlockEvent
     -> m BlockEventResult
+
 runBlockEvent (BlkEvApply ev) =
-   (BlockEventSuccess <$ verifyAndApplyBlocks' (ev ^. beaInput))
-      `catch` (return . mkBlockEventFailure ev)
+    (onSuccess <$ verifyAndApplyBlocks' (ev ^. beaInput))
+        `catch` (return . onFailure)
+  where
+    onSuccess = case ev ^. beaOutValid of
+        BlockApplySuccess -> BlockEventSuccess (IsExpected True)
+        BlockApplyFailure -> BlockEventSuccess (IsExpected False)
+    onFailure (e :: SomeException) = case ev ^. beaOutValid of
+        BlockApplySuccess -> BlockEventFailure (IsExpected False) e
+        BlockApplyFailure -> BlockEventFailure (IsExpected True) e
+
 runBlockEvent (BlkEvRollback ev) =
-   (BlockEventSuccess <$ rollbackBlocks (ev ^. berInput))
-      `catch` (return . mkBlockEventFailure ev)
+    (onSuccess <$ rollbackBlocks (ev ^. berInput))
+       `catch` (return . onFailure)
+  where
+    onSuccess = case ev ^. berOutValid of
+        BlockRollbackSuccess   -> BlockEventSuccess (IsExpected True)
+        BlockRollbackFailure _ -> BlockEventSuccess (IsExpected False)
+    onFailure (e :: SomeException) = case ev ^. berOutValid of
+        BlockRollbackSuccess -> BlockEventFailure (IsExpected False) e
+        BlockRollbackFailure brf ->
+            let
+                isExpected = case brf of
+                    BlkRbSecurityLimitExceeded
+                        | Just cfe <- fromException e
+                        , CardanoFatalError msg <- cfe
+                        , "security risk" `T.isInfixOf` msg ->
```

Comment:
Ah, I really don't like when someone parses error messages to determine the type of error :/
I find it much better to represent different types of errors with different constructors.
For example, it can be another constructor of `CardanoFatalError`.